net processing: Add ibd check before processing block for txdownloadman

sedited opened the pull request:
Calculating the rolling bloom filters for the txorphanage takes some CPU time from the scheduler thread. This can be observed for example in [this flamegraph](https://bitcoin-dev-tools.github.io/benchcoin/results/pr-172/20066462508/mainnet-default-instrumented-base-flamegraph.svg?x=920203898521&y=780), where handling the filter takes about 2.6% of total time (and most of the scheduler thread's time). 

During ibd the entries in the tx download bloom filter are just continuously rolled over and aren't consumed, since no mempool entries are created by incoming transactions from peers during ibd. The mempool does accept transactions via RPC, or the wallet at the time, however these don't interact with the orphanage and the txdownloadman, because adding anything to those is guarded by IsInitialBlockDownload() checks as well.

We're usually latching ibd to false a few blocks before catching up to the tip, so this should also not significantly degrade the performance of the filter once fully caught up.

sedited committed:
net processing: Check if we are in ibd before processing block for txdownloadman

This avoids wasting work on calculating bloom filters that aren't
consumed during ibd and continuously re-calculated as now blocks get
validated.

Also update the functional test to document that transactions would now
be requested again once out of IBD.

Co-authored-by: Lőrinc <pap.lorinc@gmail.com>

mzumsande reviewed (commented):
Concept ACK

In addition to RPC/wallet mentioned above, the mempool could also contain transactions from disk - downloaded at a time when we weren't in IBD. This is very common for nodes that aren't online 24/7 , so that they will go back into IBD at each subsequent start. But those shouldn't really be a problem either I think.

Also, this introduces a `cs_main` lock in the `scheduler` thread, which could cause some lock contention with `msghand` - not sure how important this it, but it might go well together with #32885.

sedited commented:
> Also, this introduces a cs_main lock in the scheduler thread, which could cause some lock contention with msghand - not sure how important this it, but it might go well together with https://github.com/bitcoin/bitcoin/pull/32885.

Yes, the original PR explicitly worked around this by introducing a variable to track the ibd state: https://github.com/bitcoin/bitcoin/pull/32730/files#diff-b1e19192258d83199d8adaa5ac31f067af98f63554bfdd679bd8e8073815e69dL1911-L1912 . A prior approach of that PR passes the ibd state directly in the `BlockConnected` callback. I think blocking here might be fine (which is why it opened this), but it's not great either. What do you think of drafting this for now, and we try to make progress with #32885, or do you prefer one of the other approaches?

mzumsande commented:
> What do you think of drafting this for now, and we try to make progress with https://github.com/bitcoin/bitcoin/pull/32885, or do you prefer one of the other approaches?

I think I like the approach from #32885 because it addresses the issue at its root (so that many IBD checks would benefit, not just this one). As to whether this should be drafted I'm unsure - it's probably an improvement even with the added locking due to the saved work, but given the mixed results from benchcoin maybe we'd need more testing?

maflcko commented:
> given the mixed results from benchcoin maybe we'd need more testing?

The latest runs shows a speedup in both scenarios, but the earlier run showed a slowdown in `mainnet-default-uninstrumented`, but the only difference was a rebase, so this seems to indicate that the benchcoin run itself is not stable?

(Commit 8da8ff30a57f0fe7f627d121c5e40466f1ecb0b2 was deleted by GitHub, so I am just going after the comment https://github.com/bitcoin/bitcoin/pull/34054#issuecomment-3661120464)

maflcko commented:
I am still thinking if the `std::move` patch (https://github.com/bitcoin/bitcoin/pull/34054#issuecomment-3660534822) makes sense to include here.

sedited commented:
Rebased 42dc1725b978329e0808db5e2c3a64c038c35de3 -> 62383b79498a6e6293f22790674deecfa3516d19 ([txdownloadman_ibd_check_1](https://github.com/sedited/bitcoin/tree/txdownloadman_ibd_check_1) -> [txdownloadman_ibd_check_2](https://github.com/sedited/bitcoin/tree/txdownloadman_ibd_check_2), [compare](https://github.com/sedited/bitcoin/compare/txdownloadman_ibd_check_1..txdownloadman_ibd_check_2))

* Rebased to get changes from #34253 .

l0rinc committed:
Add functional test exercising tx downloadman recently confirmed filter

This documents existing behaviour before the change in the following
commit: The bloom filter maintained by the txdownload manager tracks
recently confirmed transasctions even during ibd. If a peer sends an INV
once IBD is over it does not re-request them.

Co-authored-by: sedited <seb.kung@gmail.com>

sedited commented on test/functional/p2p_ibd_txrelay.py:
```diff
@@ -82,6 +89,15 @@ def run_test(self):
+        self.nodes[0].bumpmocktime(NONPREF_PEER_TX_DELAY)
+        peer_inver.sync_with_ping()
+        with p2p_lock:
+            assert txid not in peer_inver.getdata_requests
```
Thanks, I think I pasted another test case by mistake.

instagibbs reviewed (approved):
ACK e5f0613503b6973dbc886eba8e999f208d84853b

If a node gets an erroneous orphan during IBD this will likely speed things up as well due to EraseForBlock.

fjahr reviewed (commented):
Code review ACK e5f0613503b6973dbc886eba8e999f208d84853b
